contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,423 @@
+        token1Decimals = _token1Decimals;
+    }
+
+    /// @notice Updates the quote price LP is willing to offer tokens at
+    /// @param _P The new fixed price LP is willing to buy and sell at
+    function quote(uint256 _P) external onlyOwner {
+        require(_P >= PRICE_MIN && _P <= PRICE_MAX, "price out of range");
+        require(_P != P, "same price");
+
+        // set new P price param
+        P = _P;
+
+        emit Quote(_P);
+    }
+
+    /// @notice Updates the fee applied on swaps
+    /// @dev Effectively acts as bid/ask spread for LP
+    /// @param _fee The new swap fee
+    function setSwapFee(uint256 _fee) external onlyOwner {
+        require(_fee <= FEE_MAX, "fee > max");
+        fee = _fee;
+        emit NewSwapFee(_fee);
+    }
+
+    /// @notice Updates the admin fee applied on private pool swaps
+    /// @dev Admin fees sent to factory owner
+    /// @param _fee The new admin fee
+    function setAdminFee(uint256 _fee) external onlyFactory {
+        require(_fee <= ADMIN_FEE_MAX, "fee > max");
+        adminFee = _fee;
+        emit NewAdminFee(_fee);
+    }
+
+    /// @notice Adds liquidity to pool
+    /// @param amounts The token amounts to add in token decimals
+    /// @param deadline The deadline before which liquidity must be added
+    // TODO: test w protocol fees
+    function addLiquidity(uint256[] memory amounts, uint256 deadline)
+        external
+        onlyOwner
+        deadlineCheck(deadline)
+        hasQuote
+        returns (uint256 minted_)
+    {
+        require(amounts.length == 2, "invalid amounts");
+
+        // get current token balances in pool
+        uint256 xBal = IERC20(token0).balanceOf(address(this));
+        uint256 yBal = IERC20(token1).balanceOf(address(this));
+
+        // convert balances to wad for liquidity calcs adjusted for protocol fees
+        uint256 xWad = _amountWad(xBal - protocolFees0, true);
+        uint256 yWad = _amountWad(yBal - protocolFees1, false);
+
+        // get D balance before add liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // transfer amounts in decimals in
+        if (amounts[0] > 0) IERC20(token0).safeTransferFrom(msg.sender, address(this), amounts[0]);
+        if (amounts[1] > 0) IERC20(token1).safeTransferFrom(msg.sender, address(this), amounts[1]);
+
+        // update amounts for actual transfer amount in case of fees on transfer
+        amounts[0] = IERC20(token0).balanceOf(address(this)) - xBal;
+        amounts[1] = IERC20(token1).balanceOf(address(this)) - yBal;
+
+        // convert amounts to wad for calcs
+        uint256 amount0Wad = _amountWad(amounts[0], true);
+        uint256 amount1Wad = _amountWad(amounts[1], false);
+
+        // balances after transfer
+        xWad += amount0Wad;
+        yWad += amount1Wad;
+
+        // calc diff with new D value
+        minted_ = _D(xWad, yWad) - _d;
+        _d += minted_;
+
+        uint256[] memory fees = new uint256[](2);
```

Comment:
Unused :)

Reply:
Used in event to mimic `amm/Swap.sol`

Reply:
Ops, this was meant to be posted in `removeLiquidity()` instead.